[rostopic] Get parent topics in rostopic hz

wkentaro opened the pull request:
Parent topics means that topics to which the specified topic subscribes to
Depends on #712

peci1 committed:
Allowed saving timestamp-less messages to Cache (using rospy.Time.now() as their timestamp).

Also added a convenience `getLast` method for retrieving the last message in the cache.

wkentaro committed:
Get multiple topics for rostopic hz

wkentaro committed:
Abstract class to implement connection based transport

ros-pull-request-builder commented:
Can one of the admins verify this patch?

rhaschke committed:
roslaunch: python evaluation of expressions

- evaluation of expressions of the form $(eval ...)
- functions env(), optenv(), anon(), find(), arg() map onto corresponding $-substitutions
- implicit access to args using their name, i.e. name instead of arg('name')

wkentaro committed:
rostopic echo: Show stat when --noarr/nostr

eric-wieser committed:
Implement `__contains__`, which is a better spelling of `has`

dirk-thomas commented:
This pull request mixes several independent features which makes it unnecessary cumbersome to review. Please consider splitting it into separate PR.

dirk-thomas commented:
Never mind I just looked at #712. I will comment on that one first.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -173,40 +176,123 @@ def print_hz(self):
-def _rostopic_hz(topic, window_size=-1, filter_expr=None):
+
+def get_parent_topics(topic, parent_topics=None, impl=False):
+    """Get topics to which the specified topic subscribes to.
```
Conceptionally a topic does not subscribe to another topic. Therefore I think the current naming is problematic. I don't have a good name for it though. But the docblock need to clarify what it means.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -173,40 +176,123 @@ def print_hz(self):
+    @type parent_topics: ``NoneType`` or ``list``
+    @param parent_topics: list of parent topics which already got.
+    @type impl: ``bool``
+    @param impl: search parent topics recursively or not
```
The meaning of `impl` is unclear without consulting the docblock. The argument should be named `recursive` instead.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -173,40 +176,123 @@ def print_hz(self):
+        return parent_topics
+    for node in pub_nodes[0]:
+        subs = [t for t, l in subscriptions if node in l]
+        subs = filter(lambda x: x not in parent_topics, subs)  # get unique
```
The filter could be directly expressed in the list comprehension a line above.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -173,40 +176,123 @@ def print_hz(self):
+        if not impl:
+            return parent_topics
+        for sub in subs:
+            parent_topics = get_parent_topics(sub, parent_topics, impl=impl)
```
Before calling this function recursively it should be checked that `sub` is not alread in `parent_topcis`. While that condition was true a few lines above after other recursive calls it might not anymore.

gerkey committed:
Enabled tests and then fixed them, with much help from @dhood

dirk-thomas committed:
reenabled roswtf tests

dirk-thomas committed:
allow roswtf test to pass without rosdep being initialized (which is the case on the farm)

dirk-thomas committed:
add missing parameter to AdvertiseOptions::createAdvertiseOptions (#733)

dirk-thomas committed:
make LogAppender and Token destructor virtual (#729)

dirk-thomas committed:
change return value of param() to bool (#753)

dirk-thomas committed:
improve TopicManager::instance (#770)

eric-wieser committed:
Promote the result of read_messages to a namedtuple

dirk-thomas committed:
Merge pull request #773 from ros/log_appender_virtual_destructor

make LogAppender destructor virtual

dirk-thomas committed:
Merge pull request #774 from ros/missing_parameter

add missing parameter to AdvertiseOptions::createAdvertiseOptions

dirk-thomas committed:
Merge pull request #775 from ros/add_return_value_to_param

change return value of param() to bool

dirk-thomas committed:
Merge pull request #776 from ros/improve_topicmanager_instance

improve TopicManager::instance

dirk-thomas committed:
update changelogs

dirk-thomas committed:
1.12.0

gerkey committed:
switch to using std_msgs for tests

dhood committed:
Regression test for issue #736

codebot committed:
Use defusedxml in rosmaster

jspricke committed:
Fix confusing copyright messages/dates

Thanks to Wookey

dirk-thomas committed:
use directory specific compiler flags

gerkey committed:
Add missing inter-test dependency

gerkey committed:
Merge pull request #788 from ros/fix_test_rosbag_deps

Add missing inter-test dependency

dirk-thomas committed:
Merge pull request #782 from ros/rosmaster-defusedxml

Use defusedxml in rosmaster

dirk-thomas committed:
Merge pull request #785 from ros/directory_specific_compiler_flags

use directory specific compiler flags

dirk-thomas committed:
fix directory property

benureau committed:
Fix str conversion in encode_ros_handshake_header

The previous code was not properly filtering unicode/str
out of the header. Added the corresponding unit test.

dirk-thomas committed:
fix passing multiple args to add_rostest (fix #790)

dirk-thomas committed:
Merge pull request #791 from ros/fix_multiple_args_add_rostest

fix passing multiple args to add_rostest

dirk-thomas committed:
Merge pull request #792 from ros/fixstrconv2

fix header encoding in encode_ros_handshake_header

dirk-thomas committed:
Merge pull request #784 from ubi-agni/roslaunch-python-evaluation

python evaluation of expressions in roslaunch

rhaschke committed:
added math symbols and some standard data types for python evaluation

... and added appropriate unittests

dirk-thomas committed:
Merge pull request #793 from ubi-agni/roslaunch-math-expressions

added math symbols and some standard data types for python evaluation

dirk-thomas committed:
update changelogs

dirk-thomas committed:
1.12.1

dirk-thomas committed:
the log dir itself should not be symlinked as latest

codebot committed:
explicitly declare rosout message dependencies

dirk-thomas committed:
Merge pull request #799 from ros/add_rosgraph_msgs_dependency

explicitly declare rosout message dependencies

dirk-thomas committed:
update test to match new genpy behavior

dirk-thomas committed:
Merge pull request #803 from ros/update_test_to_new_genpy

update test to match new genpy behavior

dirk-thomas committed:
Merge pull request #780 from dhood/issue736

writing rosbags adds a newline

jasonimercer committed:
Check for pointer return from createOutgoing and fail requestTopic if it is NULL.

Commit:
Add a non verbose (quiet) mode for rosnode info

dirk-thomas committed:
Merge pull request #809 from spaghetti-/kinetic-devel

Add a non verbose (quiet) mode for rosnode info

jasonimercer committed:
Changed calling_in_this_thread restore mechanism in callOneCB from try-catch to RAII.

wkentaro committed:
Implement rospy.logXXX_throttle

Commit:
resolving naming conflicts when multiple test are added with arguments

Commit:
fix passing multiple args to roslaunch_add_file_check

Commit:
handle multiple environment arguments too

dirk-thomas committed:
Merge pull request #807 from jasonimercer/null-transport-check

Check for pointer return from createOutgoing and fail requestTopic if it is NULL.

dirk-thomas committed:
Merge pull request #811 from jasonimercer/do-not-catch-user-exception-callOneCB

Changed calling_in_this_thread restore mechanism in callOneCB from try-catch to RAII.

dirk-thomas committed:
Merge pull request #795 from ros/fix_latest_log_symlink

the log dir itself should not be symlinked as latest

dirk-thomas committed:
update changelogs

dirk-thomas committed:
1.12.2

Commit:
More detailed help string for rostopic echo -p

dirk-thomas committed:
remove extra semicolon

dirk-thomas committed:
Merge pull request #817 from ros/remove_extra_semicolon

remove extra semicolon

dirk-thomas committed:
Merge pull request #816 from spaghetti-/kinetic-devel

More detailed help string for rostopic echo -p

dirk-thomas committed:
improve debug message when queue is full (#818)

dirk-thomas committed:
Merge pull request #826 from ros/enable_tests3

Enabled tests and then fixed them

Commit:
Use logging Formatter, enabling printing exception info with exc_info=1

This recovers the funcitonality provided by the python logging library,
which includes exception information if the kwarg exc_info is provided
with a True value

Commit:
add_rostest_gtest does now add the created gtest-target as a dependeny to the created rostest

xqms committed:
rosbag_storage: make Bag constructor explicit

This catches common mistakes such as view.addQuery(bagFileName, ...)

Signed-off-by: Max Schwarz <max.schwarz@uni-bonn.de>

mikepurvis committed:
Drop strict-aliasing warnings from roslz4 file.

mikepurvis committed:
Fix to -Wno-strict-aliasing

dirk-thomas committed:
Merge pull request #839 from mikepurvis/patch-1

Drop strict-aliasing warnings from roslz4 file.

de-vri-es committed:
roslaunch: Fix <param command="..." /> for python3.

In python3, the standard output of the subprocess is of type bytes. That
causes the data to be stored as a binary xmlrpc node which ends up
base64 encoded when retrieving it later. Additionally, it crashed the
rosparam command line utility which could not deal with that.

This commit checks if the read command output is not of type `str` and
decodes it as UTF-8 in that case. This is a NOP in python2. We can't
check if the read output is of type bytes, because in python2 `bytes`
is just an alias for `str`.

dirk-thomas committed:
Merge pull request #840 from delftrobotics/roslaunch-param-command-python3

roslaunch: Fix <param command="..." /> for python3.

dirk-thomas committed:
fix compiler warnings in release mode

dirk-thomas committed:
set lz4_FOUND in order to continue using it with catkin_package(DEPENDS) (ros/catkin#813)

dirk-thomas committed:
Merge pull request #842 from ros/fix_warnings_in_release

fix compiler warnings in release mode

wkentaro committed:
Fix rostopic hz for multiple topics without no message

To fix below error

```
% rostopic hz /input_0 /input_1
subscribed to [/input_0]
subscribed to [/input_1]
no new messages
no new messages
topic      rate    min_delta   max_delta    std_dev    window
==============================================================
/input_0   9.997    0.0999      0.1002      8.891e-05   7
/input_1   0.9998   0.9995      1.001       0.0005633   7

topic      rate    min_delta   max_delta    std_dev    window
==============================================================
/input_0   9.998    0.0999      0.1002      7.342e-05   17
/input_1   0.9998   0.9995      1.001       0.0004887   17

topic      rate    min_delta   max_delta    std_dev    window
==============================================================
/input_0   9.999    0.09982     0.1002      7.967e-05   21
/input_1   0.9999   0.9995      1.001       0.0004702   21

no new messages
no new messages
no new messages
Traceback (most recent call last):
File "/home/wkentaro/ros/indigo/devel/bin/rostopic", line 6, in <module>
exec(fh.read())
File "<string>", line 35, in <module>
File "<string>", line 1977, in rostopicmain
File "<string>", line 1385, in _rostopic_cmd_hz
File "<string>", line 273, in _rostopic_hz
File "<string>", line 221, in print_hz
File "<string>", line 228, in _get_ascii_table
ValueError: max() arg is an empty sequence
```

dirk-thomas committed:
Merge pull request #843 from wkentaro/rostopic-hz-no-messages

Fix rostopic hz for multiple topics without no message

luke-gumbley committed:
Add 'Darwin' to unix-like platforms

Prevents issue calling socket.getaddrinfo with the machine name where this is not a known local hostname (e.g. where the machine name hasn't been added to the hosts file).

de-vri-es committed:
Fix `rosservice call` for python3.

dirk-thomas committed:
Merge pull request #847 from delftrobotics/rosservice-python3

Fix `rosservice call` for python3.

dirk-thomas committed:
Merge pull request #844 from jspricke/fix_years

Fix confusing copyright messages/dates

dirk-thomas committed:
Merge pull request #830 from cwecht/rostest_dependencies_bin

 add_rostest_gtest does now add the created gtest-target as a depende…

dirk-thomas committed:
Merge pull request #835 from xqms/rosbag-explicit-constructor

rosbag_storage: make Bag constructor explicit

wkentaro committed:
[message_filters] Fast approximate time synchronization in message_filters (pure python) (#802)

* Fast approximate time synchronization in message_filters

* Don't break the api of message_filters Synchronizers

* Required changes to merge origin/kinetic-devel

Commit:
update rosbag filter progress meter to use raw uncompressed input size (#857)

wkentaro committed:
List message files only for specified package

As below

```
% rosmsg list --help
Usage: rosmsg list [package_name]

Options:
  -h, --help  show this help message and exit

% rosmsg list std_msgs
std_msgs/Bool
std_msgs/Byte
std_msgs/ByteMultiArray
std_msgs/Char
std_msgs/ColorRGBA
std_msgs/Duration
std_msgs/Empty
std_msgs/Float32
std_msgs/Float32MultiArray

% rosmsg list std_msg
WARNING: no msg is found for package [std_msg]

% rosmsg list
```

dirk-thomas committed:
Revert "List message files only for specified package"

This reverts commit 4192c864160f99ea87db03b85fa9f8ac79b5e070.

contradict committed:
Allow 64-bit sizes to be passed to robag max_size (#865)

contradict committed:
Record a maximum number of splits and then begin deleting old files (#866)

* Record a maximum number of splits and then begin deleting old files

* More helpful help string

* Correct indentation

rhaschke committed:
fix unknown dependency

dirk-thomas committed:
Merge pull request #869 from ubi-agni/kinetic-devel

fix unknown msg-generation dependency

wkentaro committed:
Get parent topics (upstream in pipeline) in rostopic hz

wkentaro commented:
Moved to #874
